### DISCUSSION TITLE
Is it possible to use nanobind to embed Python in C++?

### DISCUSSION THREAD

**Discussion by cansik** [wjakob/nanobind#302] [Q&A] [answered]
We currently have a scenario where our primary application is developed in C++. However, we are interested in extending its functionality using Python, not just as a scripting language, but to create plugins that use libraries such as numpy and other dependencies.

It seems that most of the documentation for libraries such as pybind11, nanobind and Cython focuses primarily on binding C++ libraries into Python environments, with limited information on the reverse process. We have already managed to load a Python module from C++ by embedding Python via `Python.h'. We have also been able to dynamically create modules to call C++ functions from Python. However, this approach involves a significant amount of boilerplate code, so we are looking for a more streamlined solution that still meets our requirements.

It would be greatly appreciated if someone could confirm whether our desired approach is feasible and, if so, whether nanobind supports this functionality.

**Comment by wjakob** [comment, marked as answer]
Nanobind does not support this functionality, and it is (intentionally) considered out of scope to avoid feature creep (it is a "nano" project)

**Reply by thelfer** [reply]
That feature is definitively something that my project would benifit from. I am currently testing `nanobind` extensively and its works great. However, we may switch to `pybind` because of this feature.

Another option would be to build another library on top of `nanobind` to allow this. @wjakob, aside from the inherent complexity of creating such library, would it be technically feasible, i.e. can we do this non intrusively with respect to `nanobind` ?

**Reply by wjakob** [reply]
I can't say. This is one part of Python I do not know well.

**Reply by thelfer** [reply]
ok. Thanks for your answer

**Reply by btakita** [reply]
@thelfer What did you end up using? I'm considering using Nanobind + QT Quick vs Pyside to create a cross-platform app. Leaning toward Nanobind + QT if it doesn't have any future blockers. So info on something like this integration is a great indicator of how this approach supports the potential paths of growth for an app.

**Reply by thelfer** [reply]
@btakita not much I would say. I recently ported some of my codes to `pybind11` du to this restriction, knowing that transition for `nanobind` to `pybind11` is straighforward if  I ever need to do it.
